File: app/src/main/java/closer/vlllage/com/closer/pool/PoolMember.java
package closer.vlllage.com.closer.pool;

public class PoolMember {

    private Pool pool;

    PoolMember setPool(Pool pool) {
        this.pool = pool;
        onPoolInit();
        return this;
    }

    protected void onPoolInit() { }
    protected void onPoolEnd() { }

    protected <T extends PoolMember> T $(Class<T> memeber) {
        return pool.$(memeber);
    }
}

File: app/src/main/java/closer/vlllage/com/closer/pool/PoolMember.kt
package closer.vlllage.com.closer.pool

open class PoolMember {

    internal lateinit var pool: Pool

    internal fun setPool(pool: Pool): PoolMember {
        this.pool = pool
        onPoolInit()
        return this
    }

    protected open fun onPoolInit() {}
    open fun onPoolEnd() {}

    protected fun <T : PoolMember> `$`(member: Class<T>): T {
        return pool.`$`(member)
    }
}
